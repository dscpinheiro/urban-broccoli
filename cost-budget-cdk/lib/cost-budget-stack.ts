import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { aws_budgets as budgets } from 'aws-cdk-lib';

export class BudgetAppStack extends cdk.Stack {
    constructor(scope: Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props);

        // Parameters
        const thresholdAmount = new cdk.CfnParameter(this, 'BudgetThreshold', {
            type: 'Number',
            default: 50,
            minValue: 0
        });

        const notificationAddress = new cdk.CfnParameter(this, 'NotificationAddress');

        const notificationType = new cdk.CfnParameter(this, 'NotificationType', {
            type: 'String',
            default: 'EMAIL',
            allowedValues: ['EMAIL', 'SNS']
        });

        this.templateOptions.metadata = {
            'AWS::CloudFormation::Interface': {
                ParameterGroups: [
                    {
                        Label: { default: 'AWS Budget configuration' },
                        Parameters: [
                            thresholdAmount.logicalId,
                            notificationAddress.logicalId,
                            notificationType.logicalId
                        ]
                    }
                ],
                ParameterLabels: {
                    [thresholdAmount.logicalId]: {
                        default: 'Budgeted amount (in USD)'
                    },
                    [notificationAddress.logicalId]: {
                        default: 'Address that AWS sends budget notifications to, either an SNS topic or an email'
                    },
                    [notificationType.logicalId]: {
                        default: 'Type of notification that AWS sends to a subscriber'
                    }
                }
            }
        };

        // Resources
        const budgetSubscribers = [{
            address: notificationAddress.valueAsString,
            subscriptionType: notificationType.valueAsString
        }];

        const myBudget = new budgets.CfnBudget(this, 'MonthlyBudget', {
            budget: {
                budgetName: `monthly-budget-${cdk.Aws.ACCOUNT_ID}`,
                budgetType: 'COST',
                budgetLimit: {
                    amount: thresholdAmount.valueAsNumber,
                    unit: 'USD'
                },
                timeUnit: 'MONTHLY'
            },
            notificationsWithSubscribers: [
                {
                    notification: {
                        comparisonOperator: 'GREATER_THAN',
                        notificationType: 'FORECASTED',
                        threshold: 50,
                        thresholdType: 'PERCENTAGE'
                    },
                    subscribers: budgetSubscribers
                },
                {
                    notification: {
                        comparisonOperator: 'GREATER_THAN',
                        notificationType: 'FORECASTED',
                        threshold: 100,
                        thresholdType: 'PERCENTAGE'
                    },
                    subscribers: budgetSubscribers
                },

                {
                    notification: {
                        comparisonOperator: 'GREATER_THAN',
                        notificationType: 'ACTUAL',
                        threshold: 100,
                        thresholdType: 'PERCENTAGE'
                    },
                    subscribers: budgetSubscribers
                },
                {
                    notification: {
                        comparisonOperator: 'GREATER_THAN',
                        notificationType: 'ACTUAL',
                        threshold: 200,
                        thresholdType: 'PERCENTAGE'
                    },
                    subscribers: budgetSubscribers
                },
                {
                    notification: {
                        comparisonOperator: 'GREATER_THAN',
                        notificationType: 'ACTUAL',
                        threshold: 300,
                        thresholdType: 'PERCENTAGE'
                    },
                    subscribers: budgetSubscribers
                }
            ]
        });

        // Outputs
        new cdk.CfnOutput(this, 'BudgetName', { value: myBudget.ref });
    }
}
