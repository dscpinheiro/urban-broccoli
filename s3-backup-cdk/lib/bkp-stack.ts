import * as cdk from '@aws-cdk/core';
import * as s3 from '@aws-cdk/aws-s3';
import * as kms from '@aws-cdk/aws-kms';
import * as iam from '@aws-cdk/aws-iam';

export class BackupStack extends cdk.Stack {
    constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props);

        const keyAdmin = iam.User.fromUserName(this, 'BackupKeyAdmin', 'keyadmin');

        const backupKey = new kms.Key(this, 'S3BackupKey', {
            admins: [keyAdmin],
            alias: 's3-backup-key',
            description: 'Key to be used for S3 SSE-KMS',
            enabled: true
        });

        const inventoryBucket = new s3.Bucket(this, 'InventoryBucket', {
            blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
            encryption: s3.BucketEncryption.S3_MANAGED,
            lifecycleRules: [
                {
                    id: 'InventoryLifecycleRule',
                    enabled: true,
                    abortIncompleteMultipartUploadAfter: cdk.Duration.days(7),
                    expiration: cdk.Duration.days(365),
                    transitions: [
                        {
                            storageClass: s3.StorageClass.INFREQUENT_ACCESS,
                            transitionAfter: cdk.Duration.days(180)
                        }
                    ]
                }
            ]
        });

        const backupBucket = new s3.Bucket(this, 'BackupBucket', {
            blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
            bucketKeyEnabled: true,
            encryption: s3.BucketEncryption.KMS,
            encryptionKey: backupKey,
            lifecycleRules: [
                {
                    id: 'BackupLifecycleRule',
                    enabled: true,
                    abortIncompleteMultipartUploadAfter: cdk.Duration.days(7),
                    noncurrentVersionExpiration: cdk.Duration.days(90),
                    noncurrentVersionTransitions: [
                        {
                            storageClass: s3.StorageClass.ONE_ZONE_INFREQUENT_ACCESS,
                            transitionAfter: cdk.Duration.days(45)
                        }
                    ],
                    transitions: [
                        {
                            storageClass: s3.StorageClass.INFREQUENT_ACCESS,
                            transitionAfter: cdk.Duration.days(60)
                        },
                        {
                            storageClass: s3.StorageClass.GLACIER,
                            transitionAfter: cdk.Duration.days(365)
                        }
                    ]
                }
            ],
            objectOwnership: s3.ObjectOwnership.BUCKET_OWNER_PREFERRED,
            versioned: true,
            inventories: [
                {
                    inventoryId: 'backup-inventory',
                    frequency: s3.InventoryFrequency.WEEKLY,
                    includeObjectVersions: s3.InventoryObjectVersion.CURRENT,
                    enabled: true,
                    destination: {
                        bucket: inventoryBucket,
                        bucketOwner: cdk.Aws.ACCOUNT_ID
                    },
                    optionalFields: ['Size', 'LastModifiedDate', 'StorageClass', 'ETag']
                }
            ]
        });

        new cdk.CfnOutput(this, 'KmsKeyArn', { value: backupKey.keyArn });
        new cdk.CfnOutput(this, 'BackupBucketArn', { value: backupBucket.bucketArn });
        new cdk.CfnOutput(this, 'InventoryBucketArn', { value: inventoryBucket.bucketArn });
    }
}
