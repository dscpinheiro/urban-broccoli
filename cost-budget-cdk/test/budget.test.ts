import * as cdk from '@aws-cdk/core';
import { SynthUtils } from '@aws-cdk/assert';
import { BudgetAppStack } from '../lib/cost-budget-stack';

test('creates cost budget', () => {
    const app = new cdk.App();
    const stack = new BudgetAppStack(app, 'TestStack');
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
