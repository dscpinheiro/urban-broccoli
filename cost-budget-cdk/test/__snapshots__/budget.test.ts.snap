// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates cost budget 1`] = `
Object {
  "Metadata": Object {
    "AWS::CloudFormation::Interface": Object {
      "ParameterGroups": Array [
        Object {
          "Label": Object {
            "default": "AWS Budget configuration",
          },
          "Parameters": Array [
            "BudgetThreshold",
            "NotificationAddress",
            "NotificationType",
          ],
        },
      ],
      "ParameterLabels": Object {
        "BudgetThreshold": Object {
          "default": "Budgeted amount (in USD)",
        },
        "NotificationAddress": Object {
          "default": "Address that AWS sends budget notifications to, either an SNS topic or an email",
        },
        "NotificationType": Object {
          "default": "Type of notification that AWS sends to a subscriber",
        },
      },
    },
  },
  "Outputs": Object {
    "BudgetName": Object {
      "Value": Object {
        "Ref": "MonthlyBudget",
      },
    },
  },
  "Parameters": Object {
    "BudgetThreshold": Object {
      "Default": 50,
      "MinValue": 0,
      "Type": "Number",
    },
    "NotificationAddress": Object {
      "Type": "String",
    },
    "NotificationType": Object {
      "AllowedValues": Array [
        "EMAIL",
        "SNS",
      ],
      "Default": "EMAIL",
      "Type": "String",
    },
  },
  "Resources": Object {
    "MonthlyBudget": Object {
      "Properties": Object {
        "Budget": Object {
          "BudgetLimit": Object {
            "Amount": Object {
              "Ref": "BudgetThreshold",
            },
            "Unit": "USD",
          },
          "BudgetName": Object {
            "Fn::Join": Array [
              "",
              Array [
                "monthly-budget-",
                Object {
                  "Ref": "AWS::AccountId",
                },
              ],
            ],
          },
          "BudgetType": "COST",
          "TimeUnit": "MONTHLY",
        },
        "NotificationsWithSubscribers": Array [
          Object {
            "Notification": Object {
              "ComparisonOperator": "GREATER_THAN",
              "NotificationType": "FORECASTED",
              "Threshold": 50,
              "ThresholdType": "PERCENTAGE",
            },
            "Subscribers": Array [
              Object {
                "Address": Object {
                  "Ref": "NotificationAddress",
                },
                "SubscriptionType": Object {
                  "Ref": "NotificationType",
                },
              },
            ],
          },
          Object {
            "Notification": Object {
              "ComparisonOperator": "GREATER_THAN",
              "NotificationType": "FORECASTED",
              "Threshold": 100,
              "ThresholdType": "PERCENTAGE",
            },
            "Subscribers": Array [
              Object {
                "Address": Object {
                  "Ref": "NotificationAddress",
                },
                "SubscriptionType": Object {
                  "Ref": "NotificationType",
                },
              },
            ],
          },
          Object {
            "Notification": Object {
              "ComparisonOperator": "GREATER_THAN",
              "NotificationType": "ACTUAL",
              "Threshold": 100,
              "ThresholdType": "PERCENTAGE",
            },
            "Subscribers": Array [
              Object {
                "Address": Object {
                  "Ref": "NotificationAddress",
                },
                "SubscriptionType": Object {
                  "Ref": "NotificationType",
                },
              },
            ],
          },
          Object {
            "Notification": Object {
              "ComparisonOperator": "GREATER_THAN",
              "NotificationType": "ACTUAL",
              "Threshold": 200,
              "ThresholdType": "PERCENTAGE",
            },
            "Subscribers": Array [
              Object {
                "Address": Object {
                  "Ref": "NotificationAddress",
                },
                "SubscriptionType": Object {
                  "Ref": "NotificationType",
                },
              },
            ],
          },
          Object {
            "Notification": Object {
              "ComparisonOperator": "GREATER_THAN",
              "NotificationType": "ACTUAL",
              "Threshold": 300,
              "ThresholdType": "PERCENTAGE",
            },
            "Subscribers": Array [
              Object {
                "Address": Object {
                  "Ref": "NotificationAddress",
                },
                "SubscriptionType": Object {
                  "Ref": "NotificationType",
                },
              },
            ],
          },
        ],
      },
      "Type": "AWS::Budgets::Budget",
    },
  },
}
`;
